modules/javafx.base/src/main/java/javafx/beans/value/ObservableValue.java

```diff
@@ -251,4 +252,63 @@ default ObservableValue<T> orElse(T constant) {
     default <U> ObservableValue<U> flatMap(Function<? super T, ? extends ObservableValue<? extends U>> mapper) {
         return new FlatMappedBinding<>(this, mapper);
     }
+
+    /**
+     * Returns an {@code ObservableValue} that holds this value whenever the given
+     * condition evaluates to {@code true}, otherwise holds the last value when
+     * {@code condition} became {@code false}. The value is updated whenever this
+     * {@code ObservableValue} changes, unless the condition currently evaluates
+     * to {@code false}.
+     * <p>
+     * The returned {@code ObservableValue} only observes this value when the given
+     * {@code condition} evaluates to {@code true}. This allows this {@code ObservableValue}
+     * and the conditional {@code ObservableValue} to be garbage collected if neither is
+     * otherwise strongly referenced when {@code condition} becomes {@code false}.
+     * <p>
+     * A currently observed binding will observe its source, which means it will not be eligible
+     * for garbage collection while source isn't. However, using {@code when} this {@code ObservableValue}
+     * can still be eligible for garbage collection when the condition is {@code false} and the
+     * conditional itself is also eligible for garbage collection.
+     * <p>
+     * Returning {@code null} from the given condition is treated the same as
+     * returning {@code false}.
```

Comment:
I would rephrase to use the "holds" language instead of the "return":

> `{@code condition}` holding `{@code null}` is treated the same as it holding {@code false}`

But is this the behavior we want? I would consider throwing when the condition is not a `true` or `false` because it's a boolean condition. `orElse` can be used to create a condition that maps `null` to something else if the user wants to accept `null`s in the condition. I'm not sure if this is really a problem.

Reply:
I've rephrased it as suggested.

As for `null` -- I've clarified here what it currently does (so it is not left undefined), but I have no strong feelings either way. 

I mainly choose this direction because JavaFX has so far opted to treat `null` as something to be ignored or converted to a default value (for `BooleanProperty`, you can set it to `null` using the generic setter, and it would become `false`).  Since the fluent system is basically built on top of `ObjectProperty` (and thus `Object`) we can't easily exclude the possibility of it being `null` (I have to accept `ObservableValue<Boolean>` instead of `BooleanProperty`).

ie:

     listView.layoutXProperty().setValue(null);

makes no sense at all, but JavaFX accepts it and uses `0.0`.

Ouch, just noticed it always creates an exception (with expensive stacktrace) whenever you do that, despite it maybe not getting logged:

    @Override
    public void setValue(Number v) {
        if (v == null) {
            Logging.getLogger().fine("Attempt to set double property to null, using default value instead.", new NullPointerException());
            set(0.0);
        } else {
            set(v.doubleValue());
        }
    }

Reply:
Yeah, the behavior of the layout property is not the best in my opinion. Let's see what others think.